cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -273,12 +284,34 @@ private void deployApp(SystemApplication application) throws Exception {
                  application.getArtifact(), configString, NOOP_PROGRAM_TERMINATOR, null, null);
   }
 
-  private boolean isAppDeployed(ApplicationId applicationId) throws Exception {
+  // Returns true if capability applier should try to deploy this application. 2 conditions when it returns true:
+  // 1. Either the application is not deployed before.
+  // 2. If application is deployed before then the app artifact of the deployed application is not the latest one
+  //    available.
+  private boolean shouldDeployApp(ApplicationId applicationId, SystemApplication application) throws Exception {
     try {
-      applicationLifecycleService.getAppDetail(applicationId);
-      return true;
+      ApplicationDetail currAppDetail = applicationLifecycleService.getAppDetail(applicationId);
+
+      ArtifactSummary summary = application.getArtifact();
```

Comment:
put this outside the try catch block since `ApplicationNotFoundException` is thrown only at Line 293